orchagent/tunneldecaporch.cpp

```diff
@@ -20,19 +27,62 @@ extern PortsOrch*       gPortsOrch;
 extern CrmOrch*         gCrmOrch;
 extern QosOrch*         gQosOrch;
 
-TunnelDecapOrch::TunnelDecapOrch(DBConnector *db, string tableName) : Orch(db, tableName)
+TunnelDecapOrch::TunnelDecapOrch(
+    DBConnector *appDb, DBConnector *stateDb,
+    DBConnector *configDb, const vector<string> &tableNames)
+    : Orch(appDb, tableNames),
+      stateTunnelDecapTable(make_unique<Table>(stateDb, STATE_TUNNEL_DECAP_TABLE_NAME)),
+      stateTunnelDecapTermTable(make_unique<Table>(stateDb, STATE_TUNNEL_DECAP_TERM_TABLE_NAME))
 {
     SWSS_LOG_ENTER();
+
+    auto cfgSubnetDecapSubTable = new SubscriberStateTable(configDb, CFG_SUBNET_DECAP_TABLE_NAME, TableConsumable::DEFAULT_POP_BATCH_SIZE, 0);
+    deque<KeyOpFieldsValuesTuple> entries;
+    cfgSubnetDecapSubTable->pops(entries);
+    // init subnet decap config
+    for (auto &entry : entries)
+    {
+        doSubnetDecapTask(entry);
+    }
+
+    Orch::addExecutor(new Consumer(cfgSubnetDecapSubTable, this, CFG_SUBNET_DECAP_TABLE_NAME));
 }
 
-void TunnelDecapOrch::doTask(Consumer& consumer)
+void TunnelDecapOrch::doTask(Consumer &consumer)
 {
     SWSS_LOG_ENTER();
 
     if (!gPortsOrch->allPortsReady())
     {
         return;
     }
+
+    string table_name = consumer.getTableName();
+    if (table_name == APP_TUNNEL_DECAP_TABLE_NAME)
+    {
+        doDecapTunnelTask(consumer);
+    }
+    else if (table_name == APP_TUNNEL_DECAP_TERM_TABLE_NAME)
+    {
+        doDecapTunnelTermTask(consumer);
+    }
+    else if (table_name == CFG_SUBNET_DECAP_TABLE_NAME)
+    {
+        doSubnetDecapTask(consumer);
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Invalid table %s", table_name.c_str());
+    }
+
+    removeUnreferencedTunnels();
```

Comment:
It's probably not good enough to put the removal code here. Think about this, if we created a tunnel, but haven't created any tunnel term (this can happen due to the order of consuming CONFIG_DB entries), then the tunnel may be removed unexpectedly.

Reply:
The initial ref count  for a tunnel is `1`, for a tunnel with no decap term, its ref count is `1`, so it won't be removed here.